File: morning_exercises/day20.ts
class BinaryNode5 {
    constructor(
        public val: number = 0,
        public left: BinaryNode5 | null = null,
        public right: BinaryNode5 | null = null,
    ) {}

    static serialiseHelper(root: BinaryNode5 | null, vals: string[]): void {
        if (root==null) {
            vals.push("N");
            return;
        }
        vals.push(root.val.toString());

        BinaryNode5.serialiseHelper(root.left, vals);
        BinaryNode5.serialiseHelper(root.right, vals);
    }

    static serialise(root: BinaryNode5 | null): string[] {
        let vals: string[] = [];
        BinaryNode5.serialiseHelper(root, vals);

        return vals;
    }

    static deserialise(input: string[]): BinaryNode5 | null {
        const val = input.shift()!;

        if (val==="N") return null;

        const node = new BinaryNode5(parseInt(val));
        node.left = BinaryNode5.deserialise(input);
        node.right = BinaryNode5.deserialise(input);
        return node;
    }
}

{
    const root = new BinaryNode5(11);
    root.left = new BinaryNode5(2);
    root.right = new BinaryNode5(3);
    root.left.left = new BinaryNode5(4);
    root.left.right = new BinaryNode5(5);

    // console.log(BinaryNode5.serialise(root));

    console.log(BinaryNode5.serialise(BinaryNode5.deserialise(BinaryNode5.serialise(root))));
}

File: morning_exercises/day20.py
from __future__ import annotations
from collections import deque

class BinaryNode:
    def __init__(self, val=0):
        self.val=val
        self.left = None
        self.right = None

    @staticmethod
    def serialiseHelper(root: BinaryNode | None, res: list)->None:
        if root is None:
            res.append('N')
            return

        res.append(str(root.val))
        BinaryNode.serialiseHelper(root.left, res)
        BinaryNode.serialiseHelper(root.right, res)

    @staticmethod
    def serialise(root: BinaryNode | None)->list:
        res=[]
        BinaryNode.serialiseHelper(root, res)
        return res
    
    @staticmethod
    def serialiseHelperDA(root: BinaryNode | None, vals: deque)->None:
        if root is None: 
            vals.append('N')
            return

        vals.append(root.val)
        BinaryNode.serialiseHelperDA(root.left, vals)
        BinaryNode.serialiseHelperDA(root.right, vals)

    @staticmethod 
    def serialiseDA(root: BinaryNode | None)->deque:
        vals = deque()

        BinaryNode.serialiseHelperDA(root, vals)
        return vals
    
    @staticmethod
    def deserialise(res: list)->BinaryNode | None:
        val = res.pop(0)
        if val=='N': return None
        node = BinaryNode(int(val))
        node.left = BinaryNode.deserialise(res)
        node.right = BinaryNode.deserialise(res)
        return node

    @staticmethod
    def deserialise_Deque(input: deque)->BinaryNode | None:
        val = input.popleft()

        if val == 'N': return None

        node = BinaryNode(int(val))
        node.left = BinaryNode.deserialise_Deque(input)
        node.right = BinaryNode.deserialise_Deque(input)
        return node

root = BinaryNode(11)
root.left = BinaryNode(2)
root.right = BinaryNode(3)
root.left.left = BinaryNode(4)
root.left.right = BinaryNode(5)

# print(BinaryNode.serialise(root))

# print(BinaryNode.serialise(BinaryNode.deserialise(BinaryNode.serialise(root))))

print(BinaryNode.serialiseDA(root))

print(BinaryNode.serialiseDA(BinaryNode.deserialise_Deque(BinaryNode.serialiseDA(root))))
